include/z64cutscene.h

```diff
@@ -344,6 +344,42 @@ typedef enum CutsceneDestination {
     /* 0x77 */ CS_DEST_ZELDAS_COURTYARD_RECEIVE_LETTER
 } CutsceneDestination;
 
+/*
+correlation with `enum SceneLayer`:
+  CS_INDEX_MANUAL:     [SCENE_LAYER_CHILD_DAY .. SCENE_LAYER_ADULT_NIGHT]
+  CS_INDEX_SCRIPTED_*: SCENE_LAYER_CUTSCENE_FIRST
+
+correlation with `enum EntranceIndex`
+  CS_INDEX_MANUAL:     base entrance index + scene layer
+  CS_INDEX_SCRIPTED_*: base entrance index + SCENE_LAYER_CUTSCENE_FIRST + (cutscene index & 0xF)
+
+`z_select.c` naming:
+  CS_INDEX_NONE:       "Stage: night"
+  CS_INDEX_LAST:       "Stage: day"
+  CS_INDEX_SCRIPTED_*: "Stage demo 0*", between [0xFFF0 .. 0xFFFA]
+*/
+typedef enum CutsceneIndex {
+    CS_INDEX_NONE = 0x0000,
+    CS_INDEX_MANUAL = 0xFFEF,
+    CS_INDEX_SCRIPTED_0 = 0xFFF0,
+    CS_INDEX_SCRIPTED_1 = 0xFFF1,
+    CS_INDEX_SCRIPTED_2 = 0xFFF2,
+    CS_INDEX_SCRIPTED_3 = 0xFFF3,
+    CS_INDEX_SCRIPTED_4 = 0xFFF4,
+    CS_INDEX_SCRIPTED_5 = 0xFFF5,
+    CS_INDEX_SCRIPTED_6 = 0xFFF6,
+    CS_INDEX_SCRIPTED_7 = 0xFFF7,
+    CS_INDEX_SCRIPTED_8 = 0xFFF8,
+    CS_INDEX_SCRIPTED_9 = 0xFFF9,
+    CS_INDEX_SCRIPTED_A = 0xFFFA,
+    CS_INDEX_SCRIPTED_B = 0xFFFB,
+    CS_INDEX_SCRIPTED_C = 0xFFFC,
+    CS_INDEX_SCRIPTED_D = 0xFFFD,
+    CS_INDEX_SCRIPTED_E = 0xFFFE,
+    CS_INDEX_SCRIPTED_F = 0xFFFF,
```

Comment:
hmm, i do see this lines
```
if (gSaveContext.save.cutsceneIndex == CS_INDEX_SCRIPTED_D) {
    gSaveContext.save.cutsceneIndex = CS_INDEX_NONE;
}
```
and they look to me as
```
if (gSaveContext.save.cutsceneIndex == CS_INDEX_RESET) {
    gSaveContext.save.cutsceneIndex = CS_INDEX_NONE;
}
```

Reply:
The name needs to make sense in all parts of code, not just in the one code block. As I stated before, 0xFFFD triggers cutscenes after the scene has been loaded. 

For example, z_bg_toki_sword is the Master Sword, and when you pull or drop the sword it will initiate a scripted cutscene by setting the cutscene script pointer and assigning `gSaveContext.cutsceneTrigger` to 1. This causes `Cutscene_UpdateScripted` to set the cutsceneIndex to 0xFFFD, which in turn starts the processing of the script.

The purpose of the block in Play_Init is to flush the cutsceneIndex state in the event that the PlayState is destroyed before the cutscene can complete normally. This can happen if e.g. the player falls out of bounds or dies mid-cutscene.

Reply:
ah, so `CS_INDEX_TRIGGERED` might do
```
// z_demo.c:2295
void Cutscene_SetupScripted(...) {
    // if is not triggered, not idling, and not in a cutscene
    if ((gSaveContext.cutsceneTrigger != 0) && (csCtx->state == CS_STATE_IDLE) && !Player_InCsMode(play))
        // then trigger
        gSaveContext.save.cutsceneIndex = CS_INDEX_RESET;
```
because it's a "state" not a "request"

Reply:
https://github.com/zeldaret/oot/pull/2286#discussion_r1861155449

`CS_INDEX_TRIGGERED` is due to
https://github.com/zeldaret/oot/blob/279759f4815c828a1bb283c93ee62ad69fc6750b/src/code/z_play.c#L332-L334
and `z_demo.c`, who assigns it to `cutsceneIndex`. from what i can gather, it means *"current value of `cutsceneIndex` is consumed, feel free to employ it"*. so, maybe `CS_INDEX_CONSUMED` or something?

Reply:
I think the most(?) important distinction here is between values < 0xFFF0 and values >= 0xFFF0. This check appears everywhere in the code to distinguish "scripted" cutscenes from "unscripted"/"manual" cutscenes (or no cutscene).

It's true that the CS index is used to figure out which scene layer to load, but that's not the only thing it's used for, it's also used wherever this >= 0xFFF0 check happens in the code. And in fact, as mzxrules points out, setting it to a different value mid-scene can be used to trigger a "scripted" cutscene without loading a new scene layer. I think really any value >= 0xFFF0 can be used for this purpose, but I guess they used 0xFFFD for debug menus, 0xFFFD for the master sword pull, and 0xFFFF for the tower barrier.

I don't really like 0xFFFF being named for a specific cutscene, while something like 0xFFF9 (which also only used for one particular cutscene) has a generic name. I think it would be better to name all the "scripted" cutscene values 0xFFF0-0xFFFF uniformly. It would make the important distinction between "scripted" and "unscripted"/"no" cutscene clearer.

Reply:
To be clear, here's my proposal:

```c
// Values used for `cutsceneIndex` or `nextCutsceneIndex`

// Values that indicate either no cutscene, or a "manual"/"unscripted" cutscene
// (not sold on these names, but I think they should be grouped together somehow)
#define CS_INDEX_NIGHT 0x0000
#define CS_INDEX_DAY 0x8000

// Values >= 0xFFF0 indicate a "scripted" cutscene. They also control which scene layer to use on scene load, // where e.g. CS_INDEX_2 loads layer 4 + 2 = 6.
#define CS_INDEX_0 0xFFF0
#define CS_INDEX_1 0xFFF1
#define CS_INDEX_2 0xFFF2
#define CS_INDEX_3 0xFFF3
#define CS_INDEX_4 0xFFF4
#define CS_INDEX_5 0xFFF5
#define CS_INDEX_6 0xFFF6
#define CS_INDEX_7 0xFFF7
#define CS_INDEX_8 0xFFF8
#define CS_INDEX_9 0xFFF9
#define CS_INDEX_A 0xFFFA
#define CS_INDEX_B 0xFFFB
#define CS_INDEX_C 0xFFFC
#define CS_INDEX_D 0xFFFD
#define CS_INDEX_E 0xFFFE
#define CS_INDEX_F 0xFFFF

// Sentinel value used for `nextCutsceneIndex` to indicate that the `cutsceneIndex` should not be changed on scene load. Otherwise, `nextCutsceneIndex` will be copied to `cutsceneIndex`.
#define CS_INDEX_NO_NEXT_CUTSCENE 0xFFEF
```

Reply:
all in all, here's my answer: https://github.com/zeldaret/oot/pull/2286#discussion_r1863961087
- grouping together `0x0000` and `0x8000` - ok, why not, `z_select.c` and DMT destination can be somewhat a reason
- don't quite agree to name `0xFFEF` "no next", sounds a bit ambiguous
- scripted / manual division is familiar to me, but i was detered from naming them as such; not needed anyway

currently it looks like that:
https://github.com/zeldaret/oot/blob/6efea0a85fdfb44b44db6a9f456789212147f505/include/z64cutscene.h#L347-L376

we can shuffle these defines any way we want

Reply:
also not sure if `0x0000` means "no cutscene"; to me it's more like "there's no reason to change it really"

but if it does, then what does mean `0xFFFD`?
https://github.com/zeldaret/oot/blob/6efea0a85fdfb44b44db6a9f456789212147f505/src/overlays/actors/ovl_En_Sa/z_en_sa.c#L382C44-L382C52
Saria checks for this value to change her animation in cutscenes, not for `0x0000`

Reply:
Trying to consolidate the replies here (we should seriously consider moving to #oot-names-discussion to make progress faster)

I guess the important thing for me is grouping the values in that way, and ideally this grouping would be reflected in the names:

0x0000 and 0x8000: no cutscene or "manual" cutscene
0xFFF0-0xFFFF: scripted cutscenes (all named uniformly)
0xFFEF: sentinel value for `nextCutsceneIndex`

I think the way it is now (putting 0xFFFD and 0xFFEF together, and 0x8000 and 0xFFFF together) is misleading, since it obscures the important difference between values < 0xFFF0 and >= 0xFFF0. Basically everything that consumes `cutsceneIndex` (including loading scene layers) makes this distinction, so I think reading the code would be easier if we make that distinction clear from the start. And of course 0xFFEF is separate since it can't be assigned to `cutsceneIndex` directly.

The actual names don't matter to me as much, and I'll stop complaining about "next empty" since I think that discussion is a distraction. Same with 0x0000/0x8000, I don't really care what the actual names are, as long as it's clear that both mean no cutscene or unscripted cutscenes.

Reply:
I don't think Saria checking for 0xFFFD is significantly different than, say, z_horse.c checking for 0xFFF1

Reply:
ok, gotcha

Reply:
i won't however add `0xFFFB`, `0xFFFC`, `0xFFFE`, because no code even mentions them

Reply:
ok, done: https://github.com/zeldaret/oot/pull/2286/commits/e9adab6d88fe2abaf812b698589e991621244120

Reply:
Thanks, that is a bit better IMO. I still think that 0xFFFD and 0xFFFF should be named CS_INDEX_D and CS_INDEX_F like the other scripted cutscenes though, sorry for disagreeing with mzxrules here. Commenting that 0xFFFC and above don't have any corresponding scene layers in the base game is good, but the scene layer bit is only one component of the "cutscene index" system here.

@mzxrules, any thoughts?

Reply:
To me, z_select uses the cutsceneIndex variable differently from the rest of the game, and thus should have separate definitions for values 0x0000 (Night) and 0x8000 (Day) that aren't defined as CS_INDEX values. Otherwise, you will never be able to provide useful names for their values.

0xFFFD is used when executing dozens of cutscenes, the Master Sword is only just one example. Pretty much every time `cutsceneTrigger` is set to 1 or 2 it starts such a scripted cutscene. I don't think it makes sense to name this `CS_INDEX_EMPTY`

0xFFFF seems like an anomaly to me. Idk if it actually represents anything meaningful since in the two instances it appears, the first case should override it with 0xFFFD if I understand the code correctly, and the second should terminate the cutscene. 

I think it makes sense for values 0xFFF0-0xFFFA to have a distinction over 0xFFFD and 0xFFFF; values 0xFFF0-0xFFFA actually serve as index offsets for the sceneLayer system in all instances where they appear. 0xFFFD is not and cannot be used for such purposes. 0xFFFF could be, but is not actually being used for such (i.e. you could replace 0xFFFF with any value >= 0xFFF0 that isn't 0xFFFD and achieve the same effect).

Reply:
> z_select
yeah, it's a special case

> I don't think it makes sense to name this CS_INDEX_EMPTY
what does? just `CS_INDEX_D` as suggested? looks too "magical", no much better than `0xFFFD`; but if it is - all right, let's

> 0xFFFF
it does. besides, `z_demo.c` sets `CS_STATE_STOP` for them right away (Ganon's tower barrier check after each of the beam cutscenes)